jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java

```diff
@@ -99,20 +100,47 @@
         };
     private static final Iterable<ConfigurationProcessorFactory> __factoryLoader = ServiceLoader.load(ConfigurationProcessorFactory.class);
     private static final XmlParser __parser = initParser();
-    public static final Comparator<Executable> EXECUTABLE_COMPARATOR = (o1, o2) ->
+    public static final Comparator<Executable> EXECUTABLE_COMPARATOR = (e1, e2) ->
     {
-        int p1 = o1.getParameterCount();
-        int p2 = o2.getParameterCount();
-        int compare = Integer.compare(p1, p2);
-        if (compare == 0 && p1 > 0)
+        // Favour methods with less parameters
+        int count = e1.getParameterCount();
+        int compare = Integer.compare(count, e2.getParameterCount());
+        if (compare == 0 && count > 0)
         {
-            boolean a1 = o1.getParameterTypes()[p1 - 1].isArray();
-            boolean a2 = o2.getParameterTypes()[p2 - 1].isArray();
-            if (a1 && !a2)
-                compare = 1;
-            else if (!a1 && a2)
-                compare = -1;
+            Parameter[] p1 = e1.getParameters();
+            Parameter[] p2 = e2.getParameters();
+
+            // Favour methods without varargs
+            compare = Boolean.compare(p1[count - 1].isVarArgs(), p2[count - 1].isVarArgs());
+            if (compare == 0)
+            {
+                // Rank by differences in the parameters
+                for (int i = 0; i < count; i++)
+                {
+                    Class<?> t1 = p1[i].getType();
+                    Class<?> t2 = p2[i].getType();
+                    if (t1 != t2)
+                    {
+                        // Favour derived type over base type
+                        compare = Boolean.compare(t1.isAssignableFrom(t2), t2.isAssignableFrom(t1));
+                        if (compare == 0)
+                        {
+                            // favour primitive type over reference
+                            compare = Boolean.compare(t2.isPrimitive(), t1.isPrimitive());
```

Comment:
It's subtle that `t1` and `t2` are inverted here.
I would prefer `Boolean.compare(!t1.isPrimitive(), !t2.isPrimitive())`.

Reply:
that's how I wrote it initially.... switching....